server/parsec/components/postgresql/invite.py

```diff
@@ -1097,55 +1374,563 @@ async def test_dump_all_invitations(
+            )
+        )
+        assert row is not None
+        greeting_attempt_internal_id, greeting_attempt_id = row
+        greeting_attempt_id = GreetingAttemptID.from_hex(greeting_attempt_id)
+        return greeting_attempt_internal_id, greeting_attempt_id
+
+    async def greeter_join_or_cancel(
+        self,
+        conn: AsyncpgConnection,
+        greeting_attempt_internal_id: int,
+        now: DateTime,
+    ) -> bool:
+        cancelled_on = await conn.fetchval(
+            *_q_greeter_join_or_cancel(
+                greeting_attempt_internal_id=greeting_attempt_internal_id,
+                now=DateTime.now(),
+            )
+        )
+        return cancelled_on is None
+
+    async def claimer_join_or_cancel(
+        self,
+        conn: AsyncpgConnection,
+        greeting_attempt_internal_id: int,
+        now: DateTime,
+    ) -> bool:
+        cancelled_on = await conn.fetchval(
+            *_q_claimer_join_or_cancel(
+                greeting_attempt_internal_id=greeting_attempt_internal_id,
+                now=DateTime.now(),
+            )
+        )
+        return cancelled_on is None
+
+    async def new_attempt_for_greeter(
+        self,
+        conn: AsyncpgConnection,
+        organization_id: OrganizationID,
+        greeter: UserID,
+        invitation_internal_id: int,
+        now: DateTime,
+    ) -> GreetingAttemptID:
+        (
+            greeting_attempt_internal_id,
+            greeting_attempt_id,
+        ) = await self.get_active_attempt(
+            conn,
+            organization_id=organization_id,
+            greeter=greeter,
+            invitation_internal_id=invitation_internal_id,
+        )
+        is_active = await self.greeter_join_or_cancel(conn, greeting_attempt_internal_id, now)
+        while not is_active:
+            (
+                greeting_attempt_internal_id,
+                greeting_attempt_id,
+            ) = await self.get_active_attempt(
+                conn,
+                organization_id=organization_id,
+                greeter=greeter,
+                invitation_internal_id=invitation_internal_id,
+            )
+            is_active = await self.greeter_join_or_cancel(conn, greeting_attempt_internal_id, now)
+        return greeting_attempt_id
+
+    async def new_attempt_for_claimer(
+        self,
+        conn: AsyncpgConnection,
+        organization_id: OrganizationID,
+        greeter: UserID,
+        invitation_internal_id: int,
+        now: DateTime,
+    ) -> GreetingAttemptID:
+        (
+            greeting_attempt_internal_id,
+            greeting_attempt_id,
+        ) = await self.get_active_attempt(
+            conn,
+            organization_id=organization_id,
+            greeter=greeter,
+            invitation_internal_id=invitation_internal_id,
+        )
+        is_active = await self.claimer_join_or_cancel(conn, greeting_attempt_internal_id, now)
+        while not is_active:
```

Comment:
I seems weird to have a loop here: given we are in a transaction we should be able to lock the session row to handle concurrency